docs/design_notes/lift.md

```diff
@@ -0,0 +1,340 @@
+    pack passes the `repack_fn` to support this.
+    This stage is similar to stage 2 except that we only lift variables and immutable variables are ignored.
+    Immutable variables cannot be updated. Therefore, they should not be returned from the transformed function. 
+
+6. *Commit stage*
+
+    `pack` expects `fn` to return a pair where the first item will simply be returned from pack and the second item should be the repacked variables.
+    The updated variables are stored in the original/un-lifted scopes such that the mutations that happen inside the transformation survive after the transformation completes.
+
+
+### Using pack example
+
+
+A minimal example of using `pack`:
+
+```python
+from flax.core import lift
+from flax.core import Scope, init, apply, nn as core_nn
+
+def lift_id(fn, variables=True, rngs=True)
+  def wrapper(scope_fn, repack_fn, variable_groups, rng_groups, *args):
+    # normally we would first call into a JAX transform here...
+    scope = scope_fn(variable_groups, rng_groups)
+    y = fn(scope, *args)
+    out_variables = repack_fn(scope)
+    return y, out_variables
+
+  return lift.pack(wrapper, (variables,), (variables,), (rngs,))
+
+x = jnp.ones((3, 2))
+y, params = init(lift_id(core_nn.dense))(random.PRNGKey(0), x, 4)
+```
+
+NOTE that most users should not need to interact with `pack` directly.
+Please open a GitHub issue when you find a use case that is not supported yet by the existing lifted transformations.
+
+### Supported transformations
+
+| Jax Transform | Supported in Linen? | Comments |
+|-|-|-|
+| vmap | ✅ |  |
+| scan | ✅ | Carry variables cannot be initialized inside the scan body. |
+| remat | ✅ |  |
+| jit | ✅ | Current implementation might cause unnecessary recompilation. |
+| pmap | ❌ |  |
+| grad | ❌ |  |
+| xmap | ❌ |  |
+| custom_vjp | ❌ |  |
+| custom_jvp | ❌ |  |
+
+
+### Linen examples
+
+Going back to our original example, we can now use `nn.vmap` to simplify our implementation:
+
+```python
+class LinenVmap(nn.Module):
+  @nn.compact
+  def __call__(self, xs):
+    VmapMLP = nn.vmap(MLP, variable_axes={'params': 0}, split_rngs={'params': True}, in_axes=0)
+    return VmapMLP(name='mlp')(xs)
+
+variables = LinenVmap().init(random.PRNGKey(0), xs)
+print(jax.tree_map(jnp.shape, variables['params']))
+"""==>
+{
+    mlp: {
+        Dense_0: {
+            bias: (3, 4),
+            kernel: (3, 2, 4),
+        },
+        Dense_1: {
+            bias: (3, 1),
+            kernel: (3, 4, 1),
+        },
+    },
+}
+"""
+```
+
+Here we use `variable_axes={'params': 0}` to indicate that parameters are vectorized rather than shared and `split_rngs={'params': True}` means each set of parameters is initialized independently.
+
+We can also extend the example with some inner state by adding a `BatchNorm` layer:
+
+```python
+class StatefulMLP(nn.Module):
+  @nn.compact
+  def __call__(self, x):
```

Comment:
Wouldn't we always to pass in `train` when using `BatchNorm`?

Reply:
done